sig-scalability/blogs/k8s-services-scalability-issues.md

```diff
@@ -0,0 +1,123 @@
+
+#### Issue
+
+Currently while serving watches, the apiserver is deep-copying deserialized endpoints objects from etcd and serializing it once for each kube-proxy watch it serves (which is 5k watches in our scale tests).
+
+#### Relevant issues/links
+
+- [#55779](https://github.com/kubernetes/kubernetes/issues/55779#issuecomment-354452477)
+- [#48938](https://github.com/kubernetes/kubernetes/issues/48938)
+- [#58050](https://github.com/kubernetes/kubernetes/issues/58050)
+
+#### Possible Solution(s)
+
+- Avoid those duplicate serializations if/when possible ([#60067](https://github.com/kubernetes/kubernetes/pull/60067) was an attempt to do this, but it didn't seem to help too much).
+- The proper solution for this would be moving from 'endpoints' -> 'endpoint' API change (discussed in detail in below issues). As with smaller objects, the serialization load would significantly decrease.
+
+
+## Endpoints traffic is quadratic in the no. of endpoints
+
+#### Issue
+
+Endpoints object for a service contains all the individual endpoints of that service. As a result, whenever even a single pod in a service is added/updated/deleted, the whole endpoints object (which includes even the other endpoints that didn't change) is re-computed, written to storage and sent to all readers. The reason for this problem is that the current API is badly designed - not allowing for efficiently updating/reading individual endpoint changes. Explanation for quadratic behaviour:
+
+- let's say we want to do rolling upgrade of a service backends
+- whenever one pod changes, you need to recompute/update the whole endpoints object (consisting of N endpoints already)
+- with N such changes (one per each pod in the service), we'd be writing O(N^2) amount of data (because each object is of size O(N))
+
+Now if you consider watches in the picture (there's one from each kube-proxy), the situation becomes even worse as the O(N^2) size traffic gets multiplied further with no. of watches (usually equal to #nodes in the cluster). Overall, this is a serious performance drawback affecting multiple components in the control-plane (apiserver, etcd, endpoints-controller, kube-proxy). The current endpoints API makes it hard to solve this problem without introducing breaking changes.
+
+#### Relevant issues/links
+
+- [#47787](https://github.com/kubernetes/kubernetes/issues/47787)
+- [#24552](https://github.com/kubernetes/kubernetes/issues/24552)
+- [#24553](https://github.com/kubernetes/kubernetes/issues/24553)
+- [#8190](https://github.com/kubernetes/kubernetes/issues/8190)
+
+#### Possible Solution(s)
+
+- Redesign endpoints API. Some properties we may like to have for the new API:
+  - Lives in an API group
+  - Not coupled to service resource lifecycle
+  - Can track both pod and non-pod endpoints
+  - Sharded into multiple etcd key/value pairs? (potentially in its own etcd)
+  - Has a Ready property
+  - Stable sharding & ordering
```

Comment:
I still don't fully follow this one. If we decide for a singular "Endpoint" object, then we don't need any sharding at all (and ordering we get just for free, because things are ordered in API itself).

Reply:
Yeah - I feel the same. Removed it for now - we can re-add later if there's a proper reason to.